File: backend/app/services/workflow_templates/canvas_validator.py
"""Canvas state validator — Phase 4 of the Admin Visual Editor.

Validates the JSONB shape stored on `workflow_templates.canvas_state`
+ `tenant_workflow_forks.canvas_state`.

Schema (locked at Phase 4):

    {
      "version": 1,
      "trigger": {
        "trigger_type": "manual" | "event" | "scheduled" | "time_after_event" | "time_of_day",
        "trigger_config": { ... }                # opaque per trigger_type
      },
      "nodes": [
        {
          "id": "n_<slug>",                       # unique within canvas
          "type": "<node_type>",                  # must be in VALID_NODE_TYPES
          "position": { "x": <number>, "y": <number> },
          "config": { ... },                      # node-type-specific
          "label": "<optional display label>"
        },
        ...
      ],
      "edges": [
        {
          "id": "e_<slug>",                       # unique within canvas
          "source": "<node id>",                  # must reference an existing node
          "target": "<node id>",                  # must reference an existing node
          "condition": "<optional expression>",   # for branching edges
          "label": "<optional display label>"
        },
        ...
      ]
    }

Cycles are forbidden by default. To express iteration, edges may
declare `"is_iteration": true` — those edges are excluded from
the cycle check. Phase 4 doesn't render iteration loops in the
admin canvas; the flag is reserved for Phase 5+.

Optional `"containers"` overlay (container-arc Phase 1): a list of
visual grouping regions over the flat graph (nodes/edges stay the
truth). Each container is
    {
      "id": "<unique>",
      "label": "<optional>",
      "members": [ { "kind": "node" | "container", "id": "<id>" }, ... ],
      "collapsed": <bool>            # P1 ships it; P2 reads it
    }
Members use a DISCRIMINATED shape — nesting-ready. P1/P2 produce only
`kind="node"`; `kind="container"` is type-allowed but unproduced until
Phase 3. Absent `containers` is valid (back-compat).

Validation runs at WRITE time (create_template + update_template +
fork_for_tenant). Resolution at READ time assumes a valid payload.
"""

from __future__ import annotations

from typing import Any, Mapping


class CanvasValidationError(ValueError):
    """The canvas_state payload violates schema rules."""


# ─── Node type vocabulary ────────────────────────────────────────


# Composition primitives that the existing engine recognizes
# (workflow_engine.py step_type + action_type dispatch tables) +
# Phase 1 registry workflow-node names. Authoring at admin scope
# can reference any of these. Engine compatibility check happens
# when a tenant adopts the template (Phase 5+).
VALID_NODE_TYPES: tuple[str, ...] = (
    # Trigger pseudo-node (one per canvas; lives in canvas root,
    # but a `start` node can exist as the entry point)
    "start",
    "end",
    # Step types from workflow_engine.py
    "input",
    "action",
    "ai_prompt",
    "send_document",
    "playwright_action",
    "condition",
    "output",
    "notification",
    # Action types (for action nodes the type can be specialized)
    "create_record",
    "update_record",
    "open_slide_over",
    "show_confirmation",
    "send_notification",
    "send_email",
    "notify_via_contact_preference",
    "log_vault_item",
    "generate_document",
    "call_service_method",
    # Phase R-6.0a — headless Generation Focus invocation + review pause
    "invoke_generation_focus",
    "invoke_review_focus",
    # Phase 1 registry workflow-node names (canonical admin
    # vocabulary; these subsume the engine's step types when
    # rendering in the admin canvas). NOTE: the redundant
    # "generation-focus-invocation" twin was retired in focus-invocation
    # reconciliation P2 (keeper = "invoke_generation_focus").
    "send-communication",
    # Phase 4 cross-tenant workflow primitives
    "cross_tenant_order",
    "cross_tenant_request",
    "cross_tenant_acknowledgment",
    # Common composition nodes
    "decision",
    "branch",
    "parallel_split",
    "parallel_join",
    "wait",
    "schedule",
)


_REQUIRED_TOP_KEYS = {"nodes", "edges", "version"}


# ─── Validator ───────────────────────────────────────────────────


def validate_canvas_state(canvas_state: Mapping[str, Any]) -> None:
    """Raise `CanvasValidationError` if the payload is malformed."""
    if not isinstance(canvas_state, dict):
        raise CanvasValidationError(
            "canvas_state must be a mapping (got "
            f"{type(canvas_state).__name__})"
        )

    # Empty {} is permitted — it represents an empty workflow draft
    # (admin author hasn't added any nodes yet). Treated equivalent
    # to {"nodes": [], "edges": [], "version": 1}.
    if len(canvas_state) == 0:
        return

    missing = _REQUIRED_TOP_KEYS - set(canvas_state.keys())
    if missing:
        raise CanvasValidationError(
            f"canvas_state missing required keys: {sorted(missing)}"
        )

    version = canvas_state["version"]
    if not isinstance(version, int) or version < 1:
        raise CanvasValidationError(
            f"canvas_state.version must be a positive integer, got {version!r}"
        )

    nodes = canvas_state["nodes"]
    edges = canvas_state["edges"]
    if not isinstance(nodes, list):
        raise CanvasValidationError("canvas_state.nodes must be a list")
    if not isinstance(edges, list):
        raise CanvasValidationError("canvas_state.edges must be a list")

    # ── Per-node validation + id uniqueness ──────────────────
    seen_node_ids: set[str] = set()
    for idx, node in enumerate(nodes):
        if not isinstance(node, dict):
            raise CanvasValidationError(
                f"nodes[{idx}] must be a mapping"
            )
        node_id = node.get("id")
        if not isinstance(node_id, str) or not node_id:
            raise CanvasValidationError(
                f"nodes[{idx}].id must be a non-empty string"
            )
        if node_id in seen_node_ids:
            raise CanvasValidationError(
                f"nodes[{idx}].id duplicates an earlier node: {node_id!r}"
            )
        seen_node_ids.add(node_id)

        node_type = node.get("type")
        if not isinstance(node_type, str) or not node_type:
            raise CanvasValidationError(
                f"nodes[{idx}].type must be a non-empty string"
            )
        if node_type not in VALID_NODE_TYPES:
            raise CanvasValidationError(
                f"nodes[{idx}].type {node_type!r} not in canonical "
                f"VALID_NODE_TYPES set ({len(VALID_NODE_TYPES)} types). "
                f"If this is a new canonical type, add it to "
                f"`canvas_validator.VALID_NODE_TYPES` first."
            )

        position = node.get("position", {})
        if not isinstance(position, dict):
            raise CanvasValidationError(
                f"nodes[{idx}].position must be a mapping"
            )

        config = node.get("config", {})
        if not isinstance(config, dict):
            raise CanvasValidationError(
                f"nodes[{idx}].config must be a mapping"
            )

    # ── Per-edge validation + id uniqueness + ref integrity ──
    seen_edge_ids: set[str] = set()
    for idx, edge in enumerate(edges):
        if not isinstance(edge, dict):
            raise CanvasValidationError(
                f"edges[{idx}] must be a mapping"
            )
        edge_id = edge.get("id")
        if not isinstance(edge_id, str) or not edge_id:
            raise CanvasValidationError(
                f"edges[{idx}].id must be a non-empty string"
            )
        if edge_id in seen_edge_ids:
            raise CanvasValidationError(
                f"edges[{idx}].id duplicates an earlier edge: {edge_id!r}"
            )
        seen_edge_ids.add(edge_id)

        source = edge.get("source")
        target = edge.get("target")
        if not isinstance(source, str) or source not in seen_node_ids:
            raise CanvasValidationError(
                f"edges[{idx}].source {source!r} doesn't reference a "
                f"declared node id"
            )
        if not isinstance(target, str) or target not in seen_node_ids:
            raise CanvasValidationError(
                f"edges[{idx}].target {target!r} doesn't reference a "
                f"declared node id"
            )

    # ── Cycle check (excluding edges flagged as iteration) ──
    _detect_cycles(nodes, edges)

    # ── Container overlay validation (container-arc Phase 1) ──
    # `containers` is OPTIONAL — absent → nothing to check (back-compat).
    # Lockstep mirror of frontend `canvas-validator.ts` validateContainers.
    _validate_containers(canvas_state, seen_node_ids)


def _validate_containers(
    canvas_state: Mapping[str, Any], seen_node_ids: set[str]
) -> None:
    """Validate the optional `containers` overlay. No-op when absent.

    Rules: container-id uniqueness; every ``kind="node"`` member references a
    declared node; every ``kind="container"`` member references a declared
    container (Phase 3a — TWO-PASS so a forward-ref to a container declared
    LATER resolves); a node OR container is a member of AT MOST ONE parent
    container (≤1-parent — extends P1's ≤1-container to container-members); the
    container-nesting graph is ACYCLIC (Phase 3a — A⊃B⊃A rejected). Empty
    member-list is valid. Lockstep mirror of frontend
    `canvas-validator.ts` validateContainers.
    """
    if "containers" not in canvas_state:
        return
    containers = canvas_state["containers"]
    if not isinstance(containers, list):
        raise CanvasValidationError("canvas_state.containers must be a list")

    # ── Pass 1 — container-level structure + collect ALL container ids ──
    # (the full id set is needed BEFORE member-ref validation so a
    # container-member referencing a container declared LATER resolves.)
    seen_container_ids: set[str] = set()
    for idx, container in enumerate(containers):
        if not isinstance(container, dict):
            raise CanvasValidationError(f"containers[{idx}] must be a mapping")

        container_id = container.get("id")
        if not isinstance(container_id, str) or not container_id:
            raise CanvasValidationError(
                f"containers[{idx}].id must be a non-empty string"
            )
        if container_id in seen_container_ids:
            raise CanvasValidationError(
                f"containers[{idx}].id duplicates an earlier container: "
                f"{container_id!r}"
            )
        seen_container_ids.add(container_id)

        if not isinstance(container.get("collapsed"), bool):
            raise CanvasValidationError(
                f"containers[{idx}].collapsed must be a boolean"
            )

        if not isinstance(container.get("members"), list):
            raise CanvasValidationError(
                f"containers[{idx}].members must be a list"
            )

    # ── Pass 2 — member structure + ref-integrity + ≤1-parent ──
    # `member_owner` tracks BOTH kinds (a node OR container in >1 parent rejects).
    member_owner: dict[str, str] = {}
    for idx, container in enumerate(containers):
        container_id = container["id"]
        for m_idx, member in enumerate(container["members"]):
            if not isinstance(member, dict):
                raise CanvasValidationError(
                    f"containers[{idx}].members[{m_idx}] must be a mapping"
                )
            kind = member.get("kind")
            if kind not in ("node", "container"):
                raise CanvasValidationError(
                    f"containers[{idx}].members[{m_idx}].kind must be "
                    f'"node" or "container"'
                )
            member_id = member.get("id")
            if not isinstance(member_id, str) or not member_id:
                raise CanvasValidationError(
                    f"containers[{idx}].members[{m_idx}].id must be a "
                    f"non-empty string"
                )
            if kind == "node":
                if member_id not in seen_node_ids:
                    raise CanvasValidationError(
                        f"containers[{idx}].members[{m_idx}] {member_id!r} "
                        f"doesn't reference a declared node id"
                    )
            else:
                # Phase 3a — container-member ref-integrity (forward-refs
                # resolve via the full pass-1 id set). No self-membership.
                if member_id not in seen_container_ids:
                    raise CanvasValidationError(
                        f"containers[{idx}].members[{m_idx}] {member_id!r} "
                        f"doesn't reference a declared container id"
                    )
                if member_id == container_id:
                    raise CanvasValidationError(
                        f"container {container_id!r} cannot be a member of itself"
                    )
            owner = member_owner.get(member_id)
            if owner is not None:
                raise CanvasValidationError(
                    f"{member_id!r} is a member of more than one container "
                    f"({owner} and {container_id})"
                )
            member_owner[member_id] = container_id

    # ── Pass 3 — nesting must be ACYCLIC (Phase 3a) ──
    _detect_container_cycles(containers)


def _detect_container_cycles(containers: list) -> None:
    """Three-color DFS over the container-nesting graph (container → its
    ``kind="container"`` member ids). Rejects a nesting cycle (A⊃B⊃A). Mirrors
    the node-graph ``_detect_cycles`` pattern; distinct error message. Runs
    after pass 2 (all container-member refs proven), so every adjacency target
    is a declared container.
    """
    adj: dict[str, list[str]] = {}
    for container in containers:
        adj[container["id"]] = [
            m["id"] for m in container["members"] if m.get("kind") == "container"
        ]

    WHITE, GRAY, BLACK = 0, 1, 2
    color: dict[str, int] = {cid: WHITE for cid in adj}

    def visit(cid: str, path: list[str]) -> None:
        if color.get(cid) == GRAY:
            cycle = " → ".join([*path, cid])
            raise CanvasValidationError(
                f"container nesting contains a cycle: {cycle}"
            )
        if color.get(cid) == BLACK:
            return
        color[cid] = GRAY
        for child in adj.get(cid, []):
            visit(child, [*path, cid])
        color[cid] = BLACK

    for cid in list(color.keys()):
        if color[cid] == WHITE:
            visit(cid, [])


def _detect_cycles(nodes: list, edges: list) -> None:
    """DFS-based cycle detector. Edges with `is_iteration=true`
    are excluded so future iteration loops don't trip the check
    (Phase 4 doesn't render them but the schema reserves the flag).
    """
    # Build adjacency excluding iteration edges
    adj: dict[str, list[str]] = {}
    for node in nodes:
        adj.setdefault(node["id"], [])
    for edge in edges:
        if edge.get("is_iteration") is True:
            continue
        adj.setdefault(edge["source"], []).append(edge["target"])

    # DFS with three-color marking
    WHITE, GRAY, BLACK = 0, 1, 2
    color: dict[str, int] = {nid: WHITE for nid in adj}

    def visit(nid: str, path: list[str]) -> None:
        if color.get(nid) == GRAY:
            cycle = " → ".join([*path, nid])
            raise CanvasValidationError(
                f"canvas_state contains a cycle: {cycle}. To express "
                f"iteration explicitly, mark the back-edge with "
                f"is_iteration=true."
            )
        if color.get(nid) == BLACK:
            return
        color[nid] = GRAY
        for neighbor in adj.get(nid, []):
            visit(neighbor, [*path, nid])
        color[nid] = BLACK

    for nid in list(color.keys()):
        if color[nid] == WHITE:
            visit(nid, [])

File: frontend/src/lib/visual-editor/workflows/canvas-validator.ts
/**
 * Frontend canvas validator — mirror of the backend's
 * `app/services/workflow_templates/canvas_validator.py`.
 *
 * The two validators MUST agree on schema rules. Tests verify
 * cross-reference — every CanvasState valid on one side is valid
 * on the other. Used by the admin canvas editor to short-circuit
 * write requests that would 400 at the backend; UI shows the
 * validation message inline before the round-trip.
 */

import type { CanvasState } from "@/bridgeable-admin/services/workflow-templates-service"


export class CanvasValidationError extends Error {}


export const VALID_NODE_TYPES: ReadonlyArray<string> = [
  "start",
  "end",
  "input",
  "action",
  "ai_prompt",
  "send_document",
  "playwright_action",
  "condition",
  "output",
  "notification",
  "create_record",
  "update_record",
  "open_slide_over",
  "show_confirmation",
  "send_notification",
  "send_email",
  "notify_via_contact_preference",
  "log_vault_item",
  "generate_document",
  "call_service_method",
  // Phase R-6.0a — headless Generation Focus + Review Focus invocations.
  // Mirrors backend `app/services/workflow_templates/canvas_validator.py`
  // VALID_NODE_TYPES additions. Frontend authoring surfaces (the new
  // InvokeGenerationFocusConfig + InvokeReviewFocusConfig inspector
  // panes) consume these node types directly.
  "invoke_generation_focus",
  "invoke_review_focus",
  "send-communication",
  "cross_tenant_order",
  "cross_tenant_request",
  "cross_tenant_acknowledgment",
  "decision",
  "branch",
  "parallel_split",
  "parallel_join",
  "wait",
  "schedule",
] as const


/** Validate a canvas_state payload. Throws `CanvasValidationError`
 * on shape violations; no-op on valid input. Empty `{}` is
 * permitted (represents an unauthored draft). */
export function validateCanvasState(
  canvas: Partial<CanvasState> | undefined | null,
): asserts canvas is CanvasState | Record<string, never> {
  if (canvas === undefined || canvas === null) {
    throw new CanvasValidationError(
      "canvas_state must be a defined object",
    )
  }
  if (typeof canvas !== "object" || Array.isArray(canvas)) {
    throw new CanvasValidationError("canvas_state must be a mapping")
  }
  // Empty {} is valid (unauthored draft).
  const keys = Object.keys(canvas)
  if (keys.length === 0) return

  const required = ["nodes", "edges", "version"]
  const missing = required.filter((k) => !(k in canvas))
  if (missing.length > 0) {
    throw new CanvasValidationError(
      `canvas_state missing required keys: ${missing.join(", ")}`,
    )
  }

  const c = canvas as CanvasState

  if (!Number.isInteger(c.version) || c.version < 1) {
    throw new CanvasValidationError(
      `canvas_state.version must be a positive integer, got ${String(c.version)}`,
    )
  }

  if (!Array.isArray(c.nodes)) {
    throw new CanvasValidationError("canvas_state.nodes must be an array")
  }
  if (!Array.isArray(c.edges)) {
    throw new CanvasValidationError("canvas_state.edges must be an array")
  }

  const validTypes = new Set(VALID_NODE_TYPES)

  // Per-node validation + id uniqueness
  const seenNodeIds = new Set<string>()
  c.nodes.forEach((node, idx) => {
    if (typeof node !== "object" || node === null) {
      throw new CanvasValidationError(`nodes[${idx}] must be a mapping`)
    }
    if (typeof node.id !== "string" || node.id.length === 0) {
      throw new CanvasValidationError(
        `nodes[${idx}].id must be a non-empty string`,
      )
    }
    if (seenNodeIds.has(node.id)) {
      throw new CanvasValidationError(
        `nodes[${idx}].id duplicates an earlier node: ${node.id}`,
      )
    }
    seenNodeIds.add(node.id)
    if (typeof node.type !== "string" || node.type.length === 0) {
      throw new CanvasValidationError(
        `nodes[${idx}].type must be a non-empty string`,
      )
    }
    if (!validTypes.has(node.type)) {
      throw new CanvasValidationError(
        `nodes[${idx}].type "${node.type}" not in VALID_NODE_TYPES`,
      )
    }
    if (typeof node.position !== "object" || node.position === null) {
      throw new CanvasValidationError(
        `nodes[${idx}].position must be a mapping`,
      )
    }
    if (typeof node.config !== "object" || Array.isArray(node.config)) {
      throw new CanvasValidationError(
        `nodes[${idx}].config must be a mapping`,
      )
    }
  })

  // Per-edge validation + id uniqueness + reference integrity
  const seenEdgeIds = new Set<string>()
  c.edges.forEach((edge, idx) => {
    if (typeof edge !== "object" || edge === null) {
      throw new CanvasValidationError(`edges[${idx}] must be a mapping`)
    }
    if (typeof edge.id !== "string" || edge.id.length === 0) {
      throw new CanvasValidationError(
        `edges[${idx}].id must be a non-empty string`,
      )
    }
    if (seenEdgeIds.has(edge.id)) {
      throw new CanvasValidationError(
        `edges[${idx}].id duplicates an earlier edge: ${edge.id}`,
      )
    }
    seenEdgeIds.add(edge.id)
    if (typeof edge.source !== "string" || !seenNodeIds.has(edge.source)) {
      throw new CanvasValidationError(
        `edges[${idx}].source "${edge.source}" doesn't reference a declared node id`,
      )
    }
    if (typeof edge.target !== "string" || !seenNodeIds.has(edge.target)) {
      throw new CanvasValidationError(
        `edges[${idx}].target "${edge.target}" doesn't reference a declared node id`,
      )
    }
  })

  // Cycle check (excluding edges with is_iteration=true)
  detectCycles(c.nodes, c.edges)

  // Container overlay validation (container-arc Phase 1). `containers` is
  // OPTIONAL — absent → nothing to check (back-compat). Mirrors the backend
  // `canvas_validator.py` container block in lockstep.
  validateContainers(c, seenNodeIds)
}


/** Validate the optional `containers` overlay. No-op when absent.
 * Rules: container-id uniqueness; every `kind:"node"` member references a
 * declared node; every `kind:"container"` member references a declared
 * container (Phase 3a — TWO-PASS so a forward-ref to a container declared
 * LATER resolves); a node OR container is a member of AT MOST ONE parent
 * container (≤1-parent — extends P1's ≤1-container to container-members); the
 * container-nesting graph is ACYCLIC (Phase 3a — A⊃B⊃A rejected). Empty
 * member-list is valid. */
function validateContainers(
  c: CanvasState,
  seenNodeIds: Set<string>,
): void {
  const containers = c.containers
  if (containers === undefined) return
  if (!Array.isArray(containers)) {
    throw new CanvasValidationError("canvas_state.containers must be an array")
  }

  // ── Pass 1 — container-level structure + collect ALL container ids ──
  // (the full id set is needed BEFORE member-ref validation so a
  // container-member referencing a container declared LATER resolves.)
  const seenContainerIds = new Set<string>()
  containers.forEach((container, idx) => {
    if (typeof container !== "object" || container === null) {
      throw new CanvasValidationError(`containers[${idx}] must be a mapping`)
    }
    if (typeof container.id !== "string" || container.id.length === 0) {
      throw new CanvasValidationError(
        `containers[${idx}].id must be a non-empty string`,
      )
    }
    if (seenContainerIds.has(container.id)) {
      throw new CanvasValidationError(
        `containers[${idx}].id duplicates an earlier container: ${container.id}`,
      )
    }
    seenContainerIds.add(container.id)

    if (typeof container.collapsed !== "boolean") {
      throw new CanvasValidationError(
        `containers[${idx}].collapsed must be a boolean`,
      )
    }
    if (!Array.isArray(container.members)) {
      throw new CanvasValidationError(
        `containers[${idx}].members must be an array`,
      )
    }
  })

  // ── Pass 2 — member structure + ref-integrity + ≤1-parent ──
  // `memberOwner` tracks BOTH kinds (a node OR container in >1 parent rejects).
  const memberOwner = new Map<string, string>()
  containers.forEach((container, idx) => {
    container.members.forEach((member, mIdx) => {
      if (typeof member !== "object" || member === null) {
        throw new CanvasValidationError(
          `containers[${idx}].members[${mIdx}] must be a mapping`,
        )
      }
      if (member.kind !== "node" && member.kind !== "container") {
        throw new CanvasValidationError(
          `containers[${idx}].members[${mIdx}].kind must be "node" or "container"`,
        )
      }
      if (typeof member.id !== "string" || member.id.length === 0) {
        throw new CanvasValidationError(
          `containers[${idx}].members[${mIdx}].id must be a non-empty string`,
        )
      }
      if (member.kind === "node") {
        if (!seenNodeIds.has(member.id)) {
          throw new CanvasValidationError(
            `containers[${idx}].members[${mIdx}] "${member.id}" doesn't reference a declared node id`,
          )
        }
      } else {
        // Phase 3a — container-member ref-integrity (forward-refs resolve via
        // the full pass-1 id set). A container can't be a member of itself.
        if (!seenContainerIds.has(member.id)) {
          throw new CanvasValidationError(
            `containers[${idx}].members[${mIdx}] "${member.id}" doesn't reference a declared container id`,
          )
        }
        if (member.id === container.id) {
          throw new CanvasValidationError(
            `container "${container.id}" cannot be a member of itself`,
          )
        }
      }
      const owner = memberOwner.get(member.id)
      if (owner !== undefined) {
        throw new CanvasValidationError(
          `"${member.id}" is a member of more than one container (${owner} and ${container.id})`,
        )
      }
      memberOwner.set(member.id, container.id)
    })
  })

  // ── Pass 3 — nesting must be ACYCLIC (Phase 3a) ──
  detectContainerCycles(containers)
}


/** Three-color DFS over the container-nesting graph (container → its
 * `kind:"container"` member ids). Rejects a nesting cycle (A⊃B⊃A). Mirrors
 * the node-graph `detectCycles` pattern; distinct error message. Runs after
 * pass 2 (all container-member refs proven), so every adjacency target is a
 * declared container. */
function detectContainerCycles(
  containers: NonNullable<CanvasState["containers"]>,
): void {
  const adj = new Map<string, string[]>()
  for (const container of containers) {
    adj.set(
      container.id,
      container.members
        .filter((m) => m.kind === "container")
        .map((m) => m.id),
    )
  }

  const WHITE = 0
  const GRAY = 1
  const BLACK = 2
  const color = new Map<string, number>()
  for (const id of adj.keys()) color.set(id, WHITE)

  function visit(id: string, path: string[]): void {
    if (color.get(id) === GRAY) {
      throw new CanvasValidationError(
        `container nesting contains a cycle: ${[...path, id].join(" → ")}`,
      )
    }
    if (color.get(id) === BLACK) return
    color.set(id, GRAY)
    for (const child of adj.get(id) ?? []) {
      visit(child, [...path, id])
    }
    color.set(id, BLACK)
  }

  for (const id of color.keys()) {
    if (color.get(id) === WHITE) visit(id, [])
  }
}


function detectCycles(
  nodes: Array<{ id: string }>,
  edges: Array<{ source: string; target: string; is_iteration?: boolean }>,
): void {
  const adj = new Map<string, string[]>()
  for (const n of nodes) adj.set(n.id, [])
  for (const e of edges) {
    if (e.is_iteration === true) continue
    if (!adj.has(e.source)) adj.set(e.source, [])
    adj.get(e.source)!.push(e.target)
  }

  // Three-color DFS
  const WHITE = 0
  const GRAY = 1
  const BLACK = 2
  const color = new Map<string, number>()
  for (const id of adj.keys()) color.set(id, WHITE)

  function visit(id: string, path: string[]): void {
    if (color.get(id) === GRAY) {
      throw new CanvasValidationError(
        `canvas_state contains a cycle: ${[...path, id].join(" → ")}. ` +
          "To express iteration explicitly, mark the back-edge with is_iteration=true.",
      )
    }
    if (color.get(id) === BLACK) return
    color.set(id, GRAY)
    for (const neighbor of adj.get(id) ?? []) {
      visit(neighbor, [...path, id])
    }
    color.set(id, BLACK)
  }

  for (const id of color.keys()) {
    if (color.get(id) === WHITE) visit(id, [])
  }
}


/** Compute summary stats for the workflow editor's metadata strip
 * — node count, edge count, terminal-node count, branching nodes. */
export function summarizeCanvas(canvas: Partial<CanvasState>): {
  nodes: number
  edges: number
  terminalNodes: number
  branchingNodes: number
} {
  if (!canvas || !canvas.nodes || !canvas.edges) {
    return { nodes: 0, edges: 0, terminalNodes: 0, branchingNodes: 0 }
  }
  const outgoing = new Map<string, number>()
  for (const e of canvas.edges) {
    if (e.is_iteration === true) continue
    outgoing.set(e.source, (outgoing.get(e.source) ?? 0) + 1)
  }
  let terminals = 0
  let branching = 0
  for (const n of canvas.nodes) {
    const o = outgoing.get(n.id) ?? 0
    if (o === 0) terminals += 1
    if (o > 1) branching += 1
  }
  return {
    nodes: canvas.nodes.length,
    edges: canvas.edges.length,
    terminalNodes: terminals,
    branchingNodes: branching,
  }
}
